crates/core/src/factories/core_fetch.rs

```diff
@@ -0,0 +1,301 @@
+type DynTransport = Arc<Mutex<dyn Transport>>;
+
+#[derive(Debug)]
+struct CoreFetch(Inner);
+
+impl CoreFetch {
+    fn new(config: CoreFetchConfig, transport: DynTransport) -> Self {
+        let inner = Inner::new();
+        inner.create_fetch_queue(config, transport);
+        Self(inner)
+    }
+}
+
+impl Fetch for CoreFetch {
+    fn add_ops(
+        &mut self,
+        op_list: Vec<OpId>,
+        source: AgentId,
+    ) -> BoxFut<'_, K2Result<()>> {
+        Box::pin(async move {
+            // Add ops to set.
+            let ops = &mut self.0.state.lock().await.ops;
+            op_list.clone().into_iter().for_each(|op_id| {
+                ops.insert((op_id, source.clone()));
+            });
+
+            // Pass ops to fetch tasks.
+            let futures = op_list.into_iter().map(|op_id| async {
+                if let Err(err) =
+                    self.0.fetch_queue_tx.send((op_id, source.clone())).await
+                {
+                    eprintln!(
+                        "could not pass fetch request to fetch task: {err}"
+                    );
+                }
+            });
+            futures::future::join_all(futures).await;
+
+            Ok(())
+        })
+    }
+}
+
+type FetchRequest = (OpId, AgentId);
+
+#[derive(Debug)]
+struct State {
+    ops: HashSet<FetchRequest>,
+    cool_down_list: HashMap<AgentId, Instant>,
+}
+
+#[derive(Debug)]
+struct Inner {
+    state: Arc<Mutex<State>>,
+    fetch_queue_tx: Sender<FetchRequest>,
+    fetch_queue_rx: Arc<Mutex<Receiver<FetchRequest>>>,
+}
+
+impl Inner {
+    pub fn new() -> Self {
+        // Create a channel to send new ops to fetch to the tasks. This is in effect the fetch queue.
+        let (fetch_queue_tx, fetch_queue_rx) = channel::<FetchRequest>(1024);
+        let fetch_queue_rx = Arc::new(Mutex::new(fetch_queue_rx));
+
+        let ops = HashSet::new();
+        let cool_down_list = HashMap::new();
+
+        Self {
+            state: Arc::new(Mutex::new(State {
+                ops,
+                cool_down_list,
+            })),
+            fetch_queue_tx,
+            fetch_queue_rx,
+        }
+    }
+
+    pub fn create_fetch_queue(
+        &self,
+        config: CoreFetchConfig,
+        transport: DynTransport,
+    ) {
+        for _ in 0..config.parallel_request_count {
+            tokio::task::spawn(Inner::fetch_task(
+                self.fetch_queue_tx.clone(),
+                self.fetch_queue_rx.clone(),
+                transport.clone(),
+                self.state.clone(),
+                config.cool_down_interval_ms,
+            ));
+        }
+    }
+
+    async fn fetch_task(
+        fetch_request_tx: Sender<FetchRequest>,
+        fetch_request_rx: Arc<Mutex<Receiver<FetchRequest>>>,
+        transport: DynTransport,
+        state: Arc<Mutex<State>>,
+        cool_down_interval: u64,
+    ) {
+        while let Some((op_id, agent_id)) =
+            fetch_request_rx.lock().await.recv().await
```

Comment:
The lock here is immediately dropped when the `recv` returns a value, so that the other task can proceed? It looks like it should, but is that guaranteed?